src/groupeddataframe/grouping.jl

```diff
@@ -1125,3 +1139,20 @@ function DataFrame(gd::GroupedDataFrame)
     resize!(idx, doff - 1)
     parent(gd)[idx, :]
 end
+
+"""
+    groupindices(gd::GroupedDataFrame)
+
+Return a vector of group indices for each row of `parent(gd)`.
+If for some row this function returns an index `i` this means that
+a data frame `gd[i]` contains this row.
+Rows not present in any of `gd` groups have their index set to `missing`.
+"""
+groupindices(gd::GroupedDataFrame) = replace(gd.groups, 0=>missing)
```

Comment:
There are two issues here:
1. even if  the initial grouping variables do not allow missing values still `GroupedDataFrame` may not map all rows of the parent, because it can be subsetted;
2. there is an issue of type stability - upfront we do not know if we will have `missing` in the return value the result would be an union of two possible types (not that bad under Julia 1.0, but still maybe it is better to guarantee one return type). Notice that this is probably the reason why `replace` also has this behavior). It is easy to change - I would just replace `replace` with a comprehension which will use the narrowest type possible.

But I am open to change it if the majority would prefer the "type unstable" option (actually this was the reason why I considered to leave `0` instead of `missing` in the first place).

Reply:
I'm not familiar enough with the package to understand the first issue. For, the second, I was thinking that the value of `skipmissing` could be in the type of the grouped dataframe.
Just to be more precise: my preferred behavior would be that, if `groupby` is called with `skipmissing = false`, `groupindices` returns `Vector{Int}`. But if `groupby` is called with `skipmissing =true`, it returns a `Vector{Union{Missing, T})`, whether or not there are actual missing values in the grouping variable.

Reply:
This proposal exactly hits the problem number 1, so let me elaborate on it.

If you create a `GroupedDataFrame` you can subset it, e.g. assume that you have a `gd` with three groups and then write something like `gd2 = gd[2:3]` this will create a new `GroupedDataFrame` with only two groups (the first one is dropped). This means that even if `skipmissing=false` you may have have some rows that behave as-if they were skipped due to `skipmissing=true`.

[This part of code](https://github.com/JuliaData/DataFrames.jl/pull/1709/files#diff-51878bdcc3c7eb068f85c4e853b4554dR114) in this PR handles this case.

Reply:
We could add a type parameter indicating whether `gd` has been built with `skipmissing=true` or is a subset. But that would just create a type instability earlier (in the case of `skipmissing` at least), so that's not necessarily a good idea. OTOH, using a comprehension would allow inference to know that the result is a `Vector{<:Union{Int,Missing}}`, which isn't too bad (and not worse than `Vector{Union{Int,Missing}}`. I guess that's strictly superior to using `replace` (which should maybe have used the same approach, not sure).

Reply:
Actually, unfortunately, using a comprehension inference is only able to say it is an `AbstractArray`:
```
julia> g(x) = [ifelse(v == 0, missing, v) for v in x]
g (generic function with 1 method)

julia> h(x) = [v == 0 ? missing : v for v in x]
h (generic function with 1 method)

julia> x = [1,2,3]
3-element Array{Int64,1}:
 1
 2
 3

julia> @code_warntype g(x)
Body::AbstractArray
1 ─ %1 = %new(Base.Generator{Array{Int64,1},getfield(Main, Symbol("##11#12"))}, getfield(Main, Symbol("##11#12"))(), x)::Base.Generator{Array{Int64,1},getfield(Main, Symbol("##11#12"))}
│   %2 = invoke Base.collect(%1::Base.Generator{Array{Int64,1},getfield(Main, Symbol("##11#12"))})::AbstractArray
└──      return %2

julia> @code_warntype h(x)
Body::AbstractArray
1 ─ %1 = %new(Base.Generator{Array{Int64,1},getfield(Main, Symbol("##13#14"))}, getfield(Main, Symbol("##13#14"))(), x)::Base.Generator{Array{Int64,1},getfield(Main, Symbol("##13#14"))}
│   %2 = invoke Base.collect(%1::Base.Generator{Array{Int64,1},getfield(Main, Symbol("##13#14"))})::AbstractArray
└──      return %2
```
so maybe we should stick to the implementation in `replace`?

Reply:
Ah yes, that's https://github.com/JuliaLang/julia/pull/30485. Maybe we can keep `replace` for now, but file an issue on the 1.0 milestone to remember we should switch before that if possible.

Reply:
Opened https://github.com/JuliaData/DataFrames.jl/issues/1713